src/pending_activations.rs

```diff
@@ -9,42 +9,74 @@ pub struct PendingActivations {
     inner: RwLock<PaInner>,
 }
 
+// An activation is identified by it's task token / run id combo. There could potentially be
+// pending activations for the same workflow run, but with different task tokens if for example
+// the workflow times out while we are relaying a task we were issued, and we poll again
+// TODO: In theory we should never actually queue them though since we handle PAs first. Assert?
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+struct ActivationId {
+    task_token: Vec<u8>,
+    run_id: String,
+}
+
+// This could be made more memory efficient with an arena or something if ever needed
 #[derive(Default)]
 struct PaInner {
-    queue: VecDeque<WfActivation>,
-    // Keys are run ids
-    count_by_id: HashMap<String, usize>,
+    queue: VecDeque<ActivationId>,
+    activations: HashMap<ActivationId, WfActivation>,
+    count_by_run_id: HashMap<String, usize>,
 }
 
 impl PendingActivations {
     pub fn push(&self, v: WfActivation) {
         let mut inner = self.inner.write();
-        *inner
-            .count_by_id
-            .entry(v.run_id.clone())
-            .or_insert_with(|| 0) += 1;
-        inner.queue.push_back(v);
+        let wf_id = ActivationId {
+            run_id: v.run_id.clone(),
+            task_token: v.task_token.clone(),
+        };
+
+        // Check if an activation with the same ID already exists, and merge joblist if so
+        if let Some(act) = inner.activations.get_mut(&wf_id) {
+            act.jobs.extend(v.jobs);
+        } else {
+            // Incr run id count
+            *inner
+                .count_by_run_id
+                .entry(v.run_id.clone())
+                .or_insert_with(|| 0) += 1;
+            // TODO: Make better if true
+            assert!(
```

Comment:
Is it considered okay to use assert in non-test code in rust?

Reply:
Does a single WFT alway map to a single activation?

Reply:
I think using assertions in production code where you know some invariant must be upheld is a good idea. It makes the code less verbose b/c you can avoid results, while also making it super clear during testing if you violate the invariant. So, this code now actually has a few of them, most of which are totally internal (keeping maps/vec in sync) and one of which is this "no PAs with the same run id but different task tokens" which must be upheld in Core itself